### PR TITLE
Introduce SdkType context

### PR DESCRIPTION
**Pull request by joheredi**
This PR introduces a context hook that allows to get an SdkType and SdkMethod from a low-level TypeSpec type such as Type or Operation.

### CODE REVIEW

File: packages/typespec-ts/src/modular/buildCodeModel.ts [hunk 1]
```diff
@@ -754,6 +755,7 @@ function emitBasicOperation(
   rlcModels: RLCModel,
   hierarchies: string[]
 ): HrlcOperation {
+  const getSdkType = useSdkTypes()!;
```
Comment (joheredi) [inline packages/typespec-ts/src/modular/buildCodeModel.ts:758]: Changes here are an example on how to use this new context.

/cc - @qiaozha, @MaryGao

File: packages/typespec-ts/src/context/sdkTypes.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  for (const sdkModel of sdkPackage.models) {
+    if (!sdkModel.__raw) {
+      continue;
+    }
+
+    sdkTypesContext.types.set(sdkModel.__raw, sdkModel);
+  }
+
+  for (const client of sdkPackage.clients) {
+    for (const method of client.methods) {
```
Comment (qiaozha) [inline packages/typespec-ts/src/context/sdkTypes.ts:64]: is every method belong to client now?

File: packages/typespec-ts/src/context/sdkTypes.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    operations: new Map<Type, SdkMethod<SdkHttpOperation>>(),
+    types: new Map<Type, SdkType>()
+  };
+
+  for (const sdkEnum of sdkPackage.enums) {
+    if (!sdkEnum.__raw) {
+      continue;
+    }
+
+    sdkTypesContext.types.set(sdkEnum.__raw, sdkEnum);
```
Comment (qiaozha) [inline packages/typespec-ts/src/context/sdkTypes.ts:52]: if sdk packages has enums, should we also have that? for future better adoption?

File: packages/typespec-ts/src/modular/buildCodeModel.ts [hunk 4]
```diff
@@ -841,6 +843,8 @@ function emitBasicOperation(
   } else {
     bodyParameter = emitBodyParameter(context, httpOperation);
     // Flatten the body parameter if it is an anonymous model
+    const sdkOperation = getSdkType(operation);
```
Comment (MaryGao) [inline packages/typespec-ts/src/modular/buildCodeModel.ts:846]: Operation is unlikely to type. In SDKType we could split all into different types and adopt into TCGC by need. And operations could be categoried into normal, LRO, paging ect or their generation could be parameters or responses. 

So `no new debt` means if we need something from TCGC, we need to replace relevant parts.

Comment (joheredi) [reply packages/typespec-ts/src/modular/buildCodeModel.ts]: We need to find the sweetspot between meaningful refactoring but keeping the changes as scoped as possible

File: packages/typespec-ts/src/context/sdkTypes.ts (hunk 2 above)
Comment (joheredi) [reply packages/typespec-ts/src/context/sdkTypes.ts]: In TCGC sdkPackage yes, every method is under a client

File: packages/typespec-ts/src/context/sdkTypes.ts (hunk 3 above)
Comment (joheredi) [reply packages/typespec-ts/src/context/sdkTypes.ts]: We are already setting the enums here

File: packages/typespec-ts/src/modular/helpers/namingHelpers.ts [hunk 5]
```diff
@@ -77,6 +76,10 @@ export function getClassicalLayerPrefix(
   return prefix.join(separator);
 }
 
+export function isDefined<T>(thing: T | undefined | null): thing is T {
```
Comment (qiaozha) [inline packages/typespec-ts/src/modular/helpers/namingHelpers.ts:79]: do we still need this? I remember Daniel's pr introduced this.

File: common/config/rush/pnpm-lock.yaml [hunk 6]
```diff
@@ -602,6 +626,26 @@ packages:
       '@typespec/versioning': 0.57.0_@typespec+compiler@0.57.0
       change-case: 5.4.4
       pluralize: 8.0.0
+    dev: false
+
+  /@azure-tools/typespec-client-generator-core/0.44.0_uovqfdr3tffszkt2tayxtbnshy:
```
Comment (qiaozha) [inline common/config/rush/pnpm-lock.yaml:631]: why we have pnpm-lock.yaml change?

File: packages/typespec-ts/test/modularUnit/anonymousModel.spec.ts [hunk 7]
```diff
@@ -364,7 +364,8 @@ describe("anonymous model", () => {
         );
       });
 
-      it("should not flatten model if spread in the payload with required parameters", async () => {
+      // TODO: Re-enable when related work is merged
+      it.skip("should not flatten model if spread in the payload with required parameters", async () => {
```
Comment (qiaozha) [inline packages/typespec-ts/test/modularUnit/anonymousModel.spec.ts:368]: I am not quite sure why this is failing now? I didn't see we have done any specific change for it yet?

File: common/config/rush/pnpm-lock.yaml (hunk 6 above)
Comment (joheredi) [reply common/config/rush/pnpm-lock.yaml:631]: Removing these changes

File: packages/typespec-ts/test/modularUnit/anonymousModel.spec.ts (hunk 7 above)
Comment (joheredi) [reply packages/typespec-ts/test/modularUnit/anonymousModel.spec.ts:368]: yup I was testing the latest version and this was a leftover

File: packages/typespec-ts/src/modular/helpers/namingHelpers.ts (hunk 5 above)
Comment (joheredi) [reply packages/typespec-ts/src/modular/helpers/namingHelpers.ts:79]: I see it is still used in other places

Comment (joheredi) [reply packages/typespec-ts/src/modular/helpers/namingHelpers.ts:79]: such as in `getModularModelFilePath`

Comment (qiaozha) [reply packages/typespec-ts/src/modular/helpers/namingHelpers.ts:79]: Oh right, I remember I've changed similar thing in modular models pr.

**Review by qiaozha — APPROVED** [review]

**Pull request merged**